providers/fips/fipsprov.c

```diff
@@ -223,14 +239,237 @@ static int dummy_evp_call(void *provctx)
+        "afe2bc438505a59a4a40daca7a895a73db575c74c13a23ad8832957d"
+        "582d38f0a6165fb0d7e9b8799e42fd3220e332e98185a0c9429757b2"
+        "d0d02c17dbaa1ff6ed93d7e73e241eaed90caf394d2bc6570f18c81f"
+        "2be5d01a2ca99ff142b5d963f9f500325e7556f95849b3ffc7479486",
+        "be1d4596a3106bd5cb4f61c57ec5f100fb7a0c82a10b82526a97d1d9"
+        "7d98eaf6",
+        NULL
+    };
+
+    if (!hextobn(dh_p_hex, &p)
+        || !hextobn(dh_q_hex, &q)
+        || !hextobn(dh_g_hex, &g)
+        || !hextobn(dh_pub_hex, &pub)
+        || !hextobn(dh_priv_hex, &priv)
+        || !hextobn(dh_peer_pub_hex, &pub_peer)
+        || !hextobin(dh_secret_exptd_hex, &kat_secret, &kat_secret_len))
+        goto err;
+
+    ossl_param_bld_init(&bld);
+    if (!ossl_param_bld_push_BN(&bld, OSSL_PKEY_PARAM_FFC_P, p)
+        || !ossl_param_bld_push_BN(&bld, OSSL_PKEY_PARAM_FFC_Q, q)
+        || !ossl_param_bld_push_BN(&bld, OSSL_PKEY_PARAM_FFC_G, g)
+        || !ossl_param_bld_push_BN(&bld, OSSL_PKEY_PARAM_DH_PUB_KEY, pub)
+        || !ossl_param_bld_push_BN(&bld, OSSL_PKEY_PARAM_DH_PRIV_KEY, priv))
+        goto err;
+    params = ossl_param_bld_to_param(&bld);
+
+    ossl_param_bld_init(&bld);
+    if (!ossl_param_bld_push_BN(&bld, OSSL_PKEY_PARAM_FFC_P, p)
+        || !ossl_param_bld_push_BN(&bld, OSSL_PKEY_PARAM_FFC_Q, q)
+        || !ossl_param_bld_push_BN(&bld, OSSL_PKEY_PARAM_FFC_G, g)
+        || !ossl_param_bld_push_BN(&bld, OSSL_PKEY_PARAM_DH_PUB_KEY, pub_peer))
+        goto err;
+
+    params_peer = ossl_param_bld_to_param(&bld);
+    if (params == NULL || params_peer == NULL)
+        goto err;
+
+    /* Create a EVP_PKEY_CTX to load the DH keys into */
+    kactx = EVP_PKEY_CTX_new_from_name(libctx, "DH", "");
+    if (kactx == NULL)
+        goto err;
+    if (EVP_PKEY_key_fromdata_init(kactx) <= 0
+        || EVP_PKEY_fromdata(kactx, &pkey, params) <= 0)
+        goto err;
+    if (EVP_PKEY_key_fromdata_init(kactx) <= 0
+        || EVP_PKEY_fromdata(kactx, &peerkey, params_peer) <= 0)
+        goto err;
+
+    /* Create a EVP_PKEY_CTX to perform key derivation */
+    dctx = EVP_PKEY_CTX_new_from_pkey(libctx, pkey);
+    if (dctx == NULL)
+        goto err;
+
+    if (EVP_PKEY_derive_init(dctx) <= 0
+        || EVP_PKEY_derive_set_peer(dctx, peerkey) <= 0
+        || EVP_PKEY_derive(dctx, secret, &secret_len) <= 0)
+        goto err;
+
+    if (secret_len != kat_secret_len
+        || memcmp(secret, kat_secret, secret_len) != 0)
+        goto err;
+    ret = 1;
+err:
+    ossl_param_bld_free(params_peer);
+    ossl_param_bld_free(params);
+    BN_free(p);
+    BN_free(q);
+    BN_free(g);
+    BN_free(pub);
+    BN_free(priv);
+    BN_free(pub_peer);
+    OPENSSL_free(kat_secret);
+    EVP_PKEY_free(pkey);
+    EVP_PKEY_free(peerkey);
+    EVP_PKEY_CTX_free(kactx);
+    EVP_PKEY_CTX_free(dctx);
+    return ret;
+}
+#endif /* OPENSSL_NO_DH */
+
+/* TODO(3.0): To be removed */
+static int dummy_evp_call(void *provctx)
```

Comment:
Side note: I wonder how much longer we need to retain this.

Reply:
Will remove when all the self tests are added..